### ISSUE TITLE
Extras string: is there a bug not allowing to send large strings

### ISSUE THREAD

**Issue by arpank-2380** [i-pi/i-pi#51]
Hello everybody,
I have written a small python interface that manages the socket communication between i-PI and Qbox. This interface utilizes the original i-PI "DriverSocket" class and "Client" class; the latter I modified according to my needs. 
I want to send some Qbox ab initio results (eigenvalues, maximally localized Wannier functions etc) to i-PI through "extras" string. I was trying the example 
(1) dia-npt-classical and (2) dia_big-npt-classical.
In the first case size of my string was ~300 and it worked perfectly.
In the second case size of my string was ~1700 and i-PI did not
receive anything and closed the connection. I am attaching herewith the relevant portion of the ipi.log. 
[ipi.log](https://github.com/i-pi/i-pi/files/3677628/ipi.log)

Any help will be greatly appreciated.

**Comment by ceriottm** [comment]
Ciao, did you manage to trigger this problem with driver.f90? It would make it much easier to debug.

**Comment by arpank-2380** [comment]
Actually, I was working on another script that can convert the i-PI .xyz trajectory file to a full Qbox input for further calculations (like KS eigenvalues). That is done. Now, I will now start looking for the Fortran code and try to check where the problem might be to accept a larger thing. I'll update you about my findings very soon.

**Comment by arpank-2380** [comment]
Hi,
I wasn't able to reproduce the problem with driver.f90. That made me feel that there might be a problem with the Q-box interface, which for socket communication, is using the "sockets.py" module from i-PI. 
I observed that the problem is not depending on the size of the string but whether it's length is even or odd. I couldn't send any string with even length, i.e., with an odd number of bytes (somehow in my system python is adding an additional 37 bytes). 
From the gained insights, I modified the Qbox interface accordingly and now it's working.
However, with the driver.f90 I didn't have such an issue and I was able to send strings with both even and odd lengths.
Thanks

**Comment by ceriottm** [comment]
Do you think it's a problem with sockets.py or with the specifics of the
QBox interface?
Soon we'll be porting to Py3, and I noticed that one of the effects is that
encoding of the strings communicated over sockets
leads to all sorts of issues. Your help will be much appreciated at that
point!

**Comment by arpank-2380** [comment]
I don't think the problem is with the specifics of the Qbox interface. That interface is just a thin layer between i-PI and Qbox. It is correctly reading Qbox driver output file and sending it to i-PI through sockets. For sockets, I didn't do anything by myself but inherited the DriverSockets class from sockets.py.
I wasn't able to send simple strings like "TEST" (41 bytes) if it's byte-size is "odd". Then I added a blank space "TEST " (42 bytes) and it worked.
I tried to do a similar thing with the driver.f90 code by changing the length of "extra" string to 4 and 5. In both case it worked. However, I am not sure whether the C-wrapper is changing it to an "even" byte size before sending it to i-PI.   
So the origin of the problem is sockets.py, but I don't know whether it is on the receiving end, sending end or both end. 
I think the problem is not appearing for integers and floating points because they are always "even" number of bytes.

**Comment by ceriottm** [comment]
If you're happy with the current solution, it's fine with me but if you
want to get to the bottom of it, could you perhaps make a mini-driver.py
that uses sockets.py to do something super-simple, even just return zero
force and energy, and see if the problem appears? That would make it easy
for us to debug what sounds like a quirk that might be signaling something
more serious.

**Comment by arpank-2380** [comment]
Well, I have something better. I modified your Fortran driver (driver.f90) so that instead of taking inputs directly from i-PI sockets, it reads from an input file. Similarly, instead of sending outputs to i-PI socket it writes output to an XML output file. That was my guinea-pig to develop this Qbox interface. I can share this Fortran driver together with the interface.py and all associated files (like example input, output etc). 

That would be many files and directories. Where should I share them?

**Comment by ceriottm** [comment]
I;m not sure it really helps - my impression is that the problem is with
the python socket client, am I missing something?

**Comment by arpank-2380** [comment]
Hello Prof. Ceriotti,
Yes, I could at least track that the problem is originating from the getforce function of the class "Driver" contained in the Sockets.py. I can see that the program flow is stopped after receiving the message length of the extras string and creating a numpy character array. I cannot track any deeper as I am not so familiar with the inner-workings of the i-PI code. 
I had an interface for driver.f90 and that interface is built based on the Sockets.py. Anyway, that might be complicated. So I simplified it and attaching herewith (toy.py).
This toy-driver is only returning zero energy, forces along with a string.
If the string is:
self._extra="I am fine"   : It is working
However,
self._extra="I am problem" : 
Then it's not working and dropping the connection.

[mini-driver.zip](https://github.com/i-pi/i-pi/files/3726495/mini-driver.zip)

**Comment by ceriottm** [comment]
@arpank-2380    is this now fixed?

**Comment by arpank-2380** [comment]
@ceriottm : The issue raised originally wasn't a problem. I have tested that a string of any length can be sent; however one has to pad it in such a way so that the final size of the string has even number of bytes. Any odd number of bytes of data cannot be sent via sockets.py. The same remains true for python3 too, where I am sending encoded strings.
This was my observation and that's how (by padding) I fixed my interface, despite not finding any python documentation saying odd number of bytes cannot be sent/received via a python socket.
So the gist is anybody writing a python interface for their code must remember to use proper padding while sending the strings.
